modules/videoio/src/cap_gstreamer.cpp

```diff
@@ -505,116 +558,304 @@ bool GStreamerCapture::grabFrame()
+            {
+                CV_LOG_DEBUG(NULL, "videoio(MSMF): gst_app_sink_pull_sample() method is not succeeded");
+                if (!gst_app_sink_is_eos(GST_APP_SINK(sink.get())))
+                {
+                    return false;
+                }
+            }
+            gst_element_query_position(sink.get(), CV_GST_FORMAT(GST_FORMAT_TIME), &impendingVideoSampleTimeNS);
+
+            if (audioStream != -1)
+            {
+                if (!usedVideoSample)
+                {
+                    usedVideoSample.swap(impendingVideoSample);
+                    std::swap(usedVideoSampleTimeNS, impendingVideoSampleTimeNS);
+                }
+                else
+                {
+                    stopFlag = true;
+                }
+                if (impendingVideoSample)
+                {
+                    emulatedFrameNumber++;
+                }
+                videoSampleDurationNS = impendingVideoSampleTimeNS - usedVideoSampleTimeNS;
+                requiredAudioTimeNS = impendingVideoSampleTimeNS - givenAudioTimeNS;
+                givenAudioTimeNS += requiredAudioTimeNS;
+            }
+            else
+            {
+                usedVideoSample.swap(impendingVideoSample);
+                std::swap(usedVideoSampleTimeNS, impendingVideoSampleTimeNS);
+                stopFlag = true;
+                emulatedFrameNumber++;
+            }
+            returnFlag = true;
+        }
     }
+    return returnFlag;
+}
+
+bool GStreamerCapture::grabAudioFrame()
+{
+    audioSample.reset(NULL);
+
+    bool returnFlag = false;
+    gint64 audioTimeNS = bufferedAudioDurationNS;
 
-    GstAudioInfo info = {};
-    gboolean audio_info_res = gst_audio_info_from_caps(&info, frame_caps);
-    if (!audio_info_res)
+    if (bufferedAudioDurationNS > requiredAudioTimeNS)
     {
-        CV_Error(Error::StsError, "GStreamer: gst_audio_info_from_caps() is failed. Can't handle unknown layout");
+        return true;
     }
-    int bpf = GST_AUDIO_INFO_BPF(&info);
-    CV_CheckGT(bpf, 0, "");
 
-    GstStructure* structure = gst_caps_get_structure(frame_caps, 0);  // no lifetime transfer
-    if (!structure)
+    while ((!vEOS) ? audioTimeNS <= requiredAudioTimeNS : !aEOS)
     {
-        CV_LOG_ERROR(NULL, "GStreamer: Can't query 'structure'-0 from GStreamer sample");
-        return false;
+        if (gst_app_sink_is_eos(GST_APP_SINK(audiosink.get())))
+        {
+            aEOS = true;
+            if (videoStream != -1 && !vEOS)
+                returnFlag = true;
+            if (videoStream == -1)
+                audioSamplePosInSamples += chunkLengthOfBytes/((audioBitPerSample/8)*nAudioChannels);
+            break;
+        }
+        else
+        {
+            audioSample.attach(gst_app_sink_pull_sample(GST_APP_SINK(audiosink.get())));
+            if (!audioSample)
+            {
+                CV_LOG_ERROR(NULL, "videoio(MSMF): gst_app_sink_pull_sample() method is not succeeded");
```

Comment:
Why do we see this error message in tests?

```
[ RUN      ] Audio.audio/1, where GetParam() = ("test_audio.mp4", 1, 132522, 0.15, GSTREAMER)
[ INFO:0@2.343] global cap_gstreamer.cpp:1378 open OpenCV | GStreamer: /home/alalek/projects/opencv/extra/testdata/highgui/audio/test_audio.mp4
[ INFO:0@2.343] global cap_gstreamer.cpp:1411 open OpenCV | GStreamer: mode - FILE
[ERROR:0@4.890] global cap_gstreamer.cpp:700 grabAudioFrame videoio(MSMF): gst_app_sink_pull_sample() method is not succeeded
[       OK ] Audio.audio/1 (2772 ms)
```

---

Why is it an error?

Reply:
> Why is it an error?

The gst_app_sink_is_eos function does not return true in some cases until it reads an empty sample. Added a condition to track this case